exercises/handling-errors/solution/src/activities.ts

```diff
@@ -0,0 +1,76 @@
+import { Address, Bill, Distance, OrderConfirmation } from './shared';
+import { ApplicationFailure } from '@temporalio/common';
+import { log } from '@temporalio/activity';
+
+export async function getDistance(address: Address): Promise<Distance> {
+  log.info('getDistance invoked; determining distance to customer address');
+
+  // this is a simulation, which calculates a fake (but consistent)
+  // distance for a customer address based on its length. The value
+  // will therefore be different when called with different addresses,
+  // but will be the same across all invocations with the same address.
+  let kilometers: number = address.line1.length + address.line2.length - 10;
+  if (kilometers < 1) {
+    kilometers = 5;
+  }
+
+  const distance = {
+    kilometers,
+  };
+
+  log.info('getDistance complete', { distance });
+  return distance;
+}
+
+export async function sendBill(bill: Bill): Promise<OrderConfirmation> {
+  log.info('sendBill invoked', { Customer: bill.customerID, Amount: bill.amount });
+
+  let chargeAmount = bill.amount;
+
+  // This month's special offer: Get $5 off all orders over $30
+  if (bill.amount > 3000) {
+    log.info('Applying discount');
+
+    chargeAmount -= 500; // reduce amount charged by 500 cents
+  }
+
+  // reject invalid amounts before calling the payment processor
+  if (chargeAmount < 0) {
+    throw ApplicationFailure.create({
+      message: `Invalid charge amount: ${chargeAmount} (must be above zero)`,
+      details: [chargeAmount],
+      nonRetryable: true
+    });
+  }
+
+  // pretend we called a payment processing service here :-)
```

Comment:
I was going to leave a nit about the emoticon and "pretend" but I don't actually know how informal we want to be. Just calling this out as a point to consider in terms of code formality.